samples/web/content/apprtc/apprtc.py

```diff
@@ -328,77 +323,139 @@ def get_room_parameters(request, room_id, client_id, is_initiator):
-    # to match string coming in from request handlers.
-    client_map[str(client.client_id)] = client
-  return client_map
+class Client:
+  messages = []
+  is_initiator = False
+  def __init__(self, is_initiator):
+    self.is_initiator = is_initiator
+  def add_message(self, msg):
+    self.messages.append(msg)
+  def clear_messages(self):
+    self.messages = []
+  def set_initiator(self, initiator):
+    self.is_initiator = initiator
+
+class Room:
+  clients = {}
+  def add_client(self, client_id, client):
+    self.clients[client_id] = client
+  def remove_client(self, client_id):
+    del self.clients[client_id]
+  def get_occupancy(self):
+    return len(self.clients)
+  def has_client(self, client_id):
+    return client_id in self.clients
+  def get_client(self, client_id):
+    return self.clients[client_id]
+  def get_other_client(self, client_id):
+    for key, client in self.clients.items():
+      if key is not client_id:
+        return client
+    return None
+  def __str__(self):
+    return str(self.clients.keys())
+
+def get_memcache_key_for_room(host, room_id):
+  return '%s/%s' % (host, room_id)
+
+def add_client_to_room(host, room_id, client_id, is_loopback):
+  key = get_memcache_key_for_room(host, room_id)
+  memcache_client = memcache.Client()
+  error = None
+  # Compare and set retry loop.
+  while True:
+    is_initiator = None
+    messages = []
+    is_full = False
+    room_state = ''
+    room = memcache_client.gets(key)
+    if room is None:
+      # 'set' and another 'gets' are needed for CAS to work.
+      if not memcache_client.set(key, Room()):
+        logging.warning('memcache.Client.set failed for key ' + key)
+        error = 'ERROR'
+        break
+      room = memcache_client.gets(key)
+
+    occupancy = room.get_occupancy()
+    if occupancy >= 2:
+      is_full = True
+      error = 'FULL'
+      break
+
+    if occupancy == 0:
+      is_initiator = True
+      room.add_client(client_id, Client(is_initiator))
+      if is_loopback:
+        room.add_client(LOOPBACK_CLIENT_ID, Client(False))
+    else:
+      is_initiator = False
+      other_client = room.get_other_client(client_id)
+      messages = other_client.messages
+      room.add_client(client_id, Client(is_initiator))
+      other_client.clear_messages()
+    room_state = str(room)
+    if memcache_client.cas(key, room):
+      logging.info('Added client ' + client_id + ' in room ' + room_id)
+      success = True
+      break
+  return {'error': error, 'is_initiator': is_initiator,
+          'messages': messages, 'is_full': is_full, 'room_state': room_state}
+
+def remove_client_from_room(host, room_id, client_id):
+  key = get_memcache_key_for_room(host, room_id)
+  memcache_client = memcache.Client()
+  # Compare and set retry loop.
+  while True:
+    room = memcache_client.gets(key)
+    if room is None:
+      logging.warning('Unknown room: ' + room_id)
+      return {'error': 'UNKNOWN_ROOM', 'room_state': None}
```

Comment:
these error codes should probably go into constants.py (to avoid problems with misspelling, at least)